cleanlab/internal/validation.py

```diff
@@ -0,0 +1,126 @@
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+"""
+Checks to ensure valid inputs for various methods.
+"""
+
+import warnings
+import numpy as np
+import pandas as pd
+from sklearn.utils import check_X_y
+
+
+def assert_valid_inputs(X, y, pred_probs=None):
+    """Checks that X, labels, and pred_probs are correctly formatted"""
+    if not isinstance(y, (list, np.ndarray, np.generic, pd.Series, pd.DataFrame)):
+        raise TypeError("labels should be a numpy array or pandas Series.")
+    y = labels_to_array(y)
+    assert_valid_class_labels(y)
+    allow_empty_X = False if pred_probs is None else True
+    if not allow_empty_X:
+        assert_nonempty_input(X)
+        try:
+            num_examples = len(X)
+            len_supported = True
+        except:
+            len_supported = False
+        if not len_supported:
+            try:
+                num_examples = X.shape[0]
+                shape_supported = True
+            except:
+                shape_supported = False
+        if (not len_supported) and (not shape_supported):
+            raise TypeError("Data features X must support either: len(X) or X.shape[0]")
+
+        if num_examples != len(y):
+            raise ValueError(
+                f"X and labels must be same length, but X is length {num_examples} and labels is length {len(y)}."
+            )
+
+        if num_examples < 2:
+            raise ValueError("Length of X (i.e. number of examples) must be at least 2.")
+
+        assert_indexing_works(X, length_X=num_examples)
+
+    if pred_probs is not None:
+        if not isinstance(pred_probs, (np.ndarray, np.generic)):
+            raise TypeError("pred_probs must be a numpy array.")
+        if len(pred_probs) != len(y):
+            raise ValueError("pred_probs and labels must have same length.")
+        # Check for valid probabilities.
+        if (pred_probs < 0).any() or (pred_probs > 1).any():
+            raise ValueError("Values in pred_probs must be between 0 and 1.")
+        if X is not None:
+            warnings.warn("When X and pred_probs are both provided, former may be ignored.")
+
+
+def assert_valid_class_labels(y):
+    """Check that labels is zero-indexed (first label is 0) and all classes present"""
+    if y.ndim != 1:
+        raise ValueError("labels must by 1D numpy array.")
+
+    unique_classes = np.unique(y)
+    if all(unique_classes != np.arange(len(unique_classes))):
+        msg = "cleanlab requires zero-indexed labels (0,1,2,..,K-1), but in "
+        msg += "your case: np.unique(labels) = {}".format(str(unique_classes))
+        raise TypeError(msg)
+
+
+def assert_nonempty_input(X):
+    if X is None:
+        raise ValueError("Data features X cannot be None. Currently X is None.")
+
+
+def assert_indexing_works(X, idx=None, length_X=None):
+    """Ensures we can do list-based indexing into ``X`` and ``y``.
+    length_X is argument passed in since sparse matrix ``X``
```

Comment:
does this method assume X is sparse or just support both sparse and non-sparse? Worth clarifying.